compiler/rustc_mir_build/src/build/expr/into.rs

```diff
@@ -344,25 +344,63 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                     })
                     .collect();
 
-                let field_names = adt_def.variant(variant_index).fields.indices();
-
-                let fields = if let Some(FruInfo { base, field_types }) = base {
-                    let place_builder = unpack!(block = this.as_place_builder(block, *base));
-
-                    // MIR does not natively support FRU, so for each
-                    // base-supplied field, generate an operand that
-                    // reads it from the base.
-                    iter::zip(field_names, &**field_types)
-                        .map(|(n, ty)| match fields_map.get(&n) {
-                            Some(v) => v.clone(),
-                            None => {
-                                let place = place_builder.clone_project(PlaceElem::Field(n, *ty));
-                                this.consume_by_copy_or_move(place.to_place(this))
-                            }
-                        })
-                        .collect()
-                } else {
-                    field_names.filter_map(|n| fields_map.get(&n).cloned()).collect()
+                let variant = adt_def.variant(variant_index);
+                let field_names = variant.fields.indices();
+
+                let fields = match base {
+                    AdtExprBase::None => {
+                        field_names.filter_map(|n| fields_map.get(&n).cloned()).collect()
+                    }
+                    AdtExprBase::Base(FruInfo { base, field_types }) => {
+                        let place_builder = unpack!(block = this.as_place_builder(block, *base));
+
+                        // MIR does not natively support FRU, so for each
+                        // base-supplied field, generate an operand that
+                        // reads it from the base.
+                        iter::zip(field_names, &**field_types)
+                            .map(|(n, ty)| match fields_map.get(&n) {
+                                Some(v) => v.clone(),
+                                None => {
+                                    let place =
+                                        place_builder.clone_project(PlaceElem::Field(n, *ty));
+                                    this.consume_by_copy_or_move(place.to_place(this))
+                                }
+                            })
+                            .collect()
+                    }
+                    AdtExprBase::DefaultFields(field_types) => {
+                        iter::zip(field_names, &**field_types)
+                            .map(|(n, ty)| match fields_map.get(&n) {
+                                Some(v) => v.clone(),
+                                None => match variant.fields[n].value {
+                                    Some(def) => {
+                                        let value =
+                                            this.tcx.instantiate_and_normalize_erasing_regions(
```

Comment:
Ended up using `Const::from_unevaluated(this.tcx, def).instantiate(this.tcx, args);`, as `Const::new_unevaluated` doesn't exist, and with the prior comment it felt like it might be correct?